core/src/main/java/org/elasticsearch/common/geo/parsers/GeoWKTParser.java

```diff
@@ -0,0 +1,279 @@
+    }
+
+    private static PointBuilder parsePoint(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        if (nextEmptyOrOpen(stream).equals(EMPTY)) {
+            return null;
+        }
+        PointBuilder pt = new PointBuilder(nextNumber(stream), nextNumber(stream));
+        if (isNumberNext(stream) == true) {
+            nextNumber(stream);
+        }
+        nextCloser(stream);
+        return pt;
+    }
+
+    private static List<Coordinate> parseCoordinates(StreamTokenizer stream)
+            throws IOException, ElasticsearchParseException {
+        CoordinatesBuilder coordinates = new CoordinatesBuilder();
+        coordinates.coordinate(parseCoordinate(stream));
+        while (nextCloserOrComma(stream).equals(COMMA)) {
+            coordinates.coordinate(parseCoordinate(stream));
+        }
+        return coordinates.build();
+    }
+
+    private static Coordinate parseCoordinate(StreamTokenizer stream)
+            throws IOException, ElasticsearchParseException {
+        final double lon = nextNumber(stream);
+        final double lat = nextNumber(stream);
+        Double z = null;
+        if (isNumberNext(stream)) {
+            z = nextNumber(stream);
+        }
+        return z == null ? new Coordinate(lon, lat) : new Coordinate(lon, lat, z);
+    }
+
+    private static MultiPointBuilder parseMultiPoint(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        String token = nextEmptyOrOpen(stream);
+        if (token.equals(EMPTY)) {
+            return null;
+        }
+        return new MultiPointBuilder(parseCoordinates(stream));
+    }
+
+    private static LineStringBuilder parseLine(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        String token = nextEmptyOrOpen(stream);
+        if (token.equals(EMPTY)) {
+            return null;
+        }
+        return new LineStringBuilder(parseCoordinates(stream));
+    }
+
+    private static MultiLineStringBuilder parseMultiLine(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        String token = nextEmptyOrOpen(stream);
+        if (token.equals(EMPTY)) {
+            return null;
+        }
+        MultiLineStringBuilder builder = new MultiLineStringBuilder();
+        builder.linestring(parseLine(stream));
+        while (nextCloserOrComma(stream).equals(COMMA)) {
+            builder.linestring(parseLine(stream));
+        }
+        return builder;
+    }
+
+    private static PolygonBuilder parsePolygon(StreamTokenizer stream) throws IOException, ElasticsearchParseException {
+        if (nextEmptyOrOpen(stream).equals(EMPTY)) {
+            return null;
+        }
+        // todo orientation?
+        PolygonBuilder builder = new PolygonBuilder(parseLine(stream), ShapeBuilder.Orientation.RIGHT);
+        if (nextWord(stream).equals(LPAREN)) {
+            while (nextCloserOrComma(stream).equals(COMMA)) {
```

Comment:
same problem as above I think, only that unmatched closing brackets will create holes here.